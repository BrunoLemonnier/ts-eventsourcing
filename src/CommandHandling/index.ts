export * from './Command';
export * from './CommandBus';
export * from './CommandHandler';
export * from './HandleCommand';
export * from './SimpleCommandBus';

// Error
export * from './Error';
