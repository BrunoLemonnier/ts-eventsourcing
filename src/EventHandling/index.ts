export * from './DomainEventBus/AsynchronousDomainEventBus';
export * from './DomainEventBus';
export * from './EventListener';
export * from './HandleDomainEvent';

// Error
export * from './Error/IncorrectDomainEventHandlerError';
