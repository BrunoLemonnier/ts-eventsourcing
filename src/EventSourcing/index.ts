export * from './EventSourcedAggregateFactory';
export * from './AggregateHandleEvent';
export * from './EventSourcedAggregateRoot';
export * from './EventSourcedEntity';
export * from './EventSourcingRepositoryInterface';

// Factory
export * from './Factory/SimpleEventSourcedAggregateFactory';

// Repository
export * from './Repository/CachedEventSourcingRepositoryDecorator';
export * from './Repository/EventSourcingRepository';

// Error
export * from './Error/IncorrectEventHandlerError';
