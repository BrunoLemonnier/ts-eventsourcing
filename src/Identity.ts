import * as uuid from 'uuid/v4';

export class Identity {

  public static create() {
    return new this(uuid());
  }

  public static of(id: Identity) {
    if (id instanceof this) {
      return id;
    }
    return new this((id as any).toString());
  }

  constructor(private readonly id: string) {

  }

  public toString() {
    return this.id;
  }

  public equals(id: Identity) {
    return this.toString() === id.toString();
  }

}
