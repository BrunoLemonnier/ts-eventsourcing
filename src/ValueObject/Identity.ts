export interface Identity {

  toString(): string;

  equals(id: Identity): boolean;

}
