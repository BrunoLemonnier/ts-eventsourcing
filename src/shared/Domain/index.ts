export * from './DomainMessage';
export * from './DomainEventStream';
export * from './DomainEventStreamDecorator';
export * from './DomainMessage';
export * from './DomainEvent';
export * from './SimpleDomainEventStream';

export * from './Decorator';

