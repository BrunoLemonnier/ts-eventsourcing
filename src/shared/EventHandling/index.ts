export * from './AsynchronousDomainEventBus';
export * from './DomainEventBus';
export * from './EventListener';
export * from './HandleDomainEvent';

// Error
export * from './Error/IncorrectDomainEventHandlerError';
