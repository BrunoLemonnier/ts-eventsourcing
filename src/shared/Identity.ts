import * as uuid from 'uuid/v4';

export class Identity {

  public static create() {
    return new this(uuid());
  }

  public static of(id: Identity) {
    return new this(id.toString());
  }

  constructor(private readonly id: string) {

  }

  public toString() {
    return this.id;
  }

  public equals(id: Identity) {
    return this.toString() === id.toString();
  }

}
