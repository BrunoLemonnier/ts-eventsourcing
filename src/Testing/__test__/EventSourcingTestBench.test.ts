/* tslint:disable:max-classes-per-file */

import { EventSourcingTestBench } from '../EventSourcingTestBench';
import { Command, CommandHandler } from '../../CommandHandling';
import { EventListener } from '../../EventHandling';
import { Identity } from '../../Identity';
import { EventSourcedAggregateRoot } from '../../EventSourcing';
import { DomainEvent, DomainEventStream, DomainMessage, SimpleDomainEventStream } from '../../Domain';

describe('givenCommandHandler should register commandHandler to the command bus', () => {
  class TestCommandHandler implements CommandHandler {

  }

  it('by value', () => {
    const testBench = new EventSourcingTestBench();
    const spy = jest.spyOn(testBench.commandBus, 'subscribe');
    const commandHandler = new TestCommandHandler();
    testBench.givenCommandHandler(commandHandler);
    expect(spy).toBeCalledWith(commandHandler);
  });

  it('by callback', () => {
    const testBench = new EventSourcingTestBench();
    const spy = jest.spyOn(testBench.commandBus, 'subscribe');
    const commandHandler = new TestCommandHandler();
    testBench.givenCommandHandler(() => commandHandler);
    expect(spy).toBeCalledWith(commandHandler);
  });

});

describe('givenEventListener should be registered to the event bus', () => {
  it('by value', () => {
    const testBench = new EventSourcingTestBench();

    class TestEventListener implements EventListener {

    }

    const spy = jest.spyOn(testBench.eventBus, 'subscribe');
    const eventListener = new TestEventListener();
    testBench.givenEventListener(eventListener);
    expect(spy).toBeCalledWith(eventListener);
  });

  it('by callback', () => {
    const testBench = new EventSourcingTestBench();

    class TestEventListener implements EventListener {

    }

    const spy = jest.spyOn(testBench.eventBus, 'subscribe');
    const eventListener = new TestEventListener();
    testBench.givenEventListener(() => eventListener);
    expect(spy).toBeCalledWith(eventListener);
  });
});

describe('should call spy factory', () => {
  // :) spy to check if can apply spies :).
  it('sync', async () => {
    const spy = jest.fn();
    const testBench = new EventSourcingTestBench();
    await testBench.givenSpies(spy);
    expect(spy).toBeCalledWith(testBench);
  });

  it('async', async () => {
    const spy = jest.fn();
    const testBench = new EventSourcingTestBench();
    await testBench.givenSpies(async (testBenchArg) => {
      spy(testBenchArg);
    });
    expect(spy).toBeCalledWith(testBench);
  });
});

describe('Given domain events should be assigned to corresponding aggregate stores', () => {
  class TestEvent implements DomainEvent {

  }

  class TestEvent2 implements DomainEvent {

  }

  class TestAggregate extends EventSourcedAggregateRoot {

  }

  it('Single instance', async () => {
    const testBench = new EventSourcingTestBench();

    const id = Identity.create();

    await testBench
      .given(id, TestAggregate, [
        new TestEvent(),
        new TestEvent2(),
      ]);

    const context = testBench.getAggregateTestContext(TestAggregate);
    const stream = await context.getEventStore().loadAll();
    const messages = await stream.toArray().toPromise();

    expect(messages).toEqual([
      new DomainMessage(id, 0, new TestEvent(), testBench.getCurrentTime()),
      new DomainMessage(id, 1, new TestEvent2(), testBench.getCurrentTime()),
    ]);
  });

  it('Multiple instances', async () => {
    const testBench = new EventSourcingTestBench();

    const id1 = Identity.create();
    const id2 = Identity.create();

    await testBench
      .given(id1, TestAggregate, [
        new TestEvent(),
        new TestEvent2(),
      ])
      .given(id2, TestAggregate, [
        new TestEvent(),
        new TestEvent2(),
        new TestEvent2(),
        new TestEvent(),
      ]);

    const context = testBench.getAggregateTestContext(TestAggregate);
    const stream = await context.getEventStore().loadAll();
    const messages = await stream.toArray().toPromise();

    expect(messages).toEqual([
      new DomainMessage(id1, 0, new TestEvent(), testBench.getCurrentTime()),
      new DomainMessage(id1, 1, new TestEvent2(), testBench.getCurrentTime()),
      new DomainMessage(id2, 0, new TestEvent(), testBench.getCurrentTime()),
      new DomainMessage(id2, 1, new TestEvent2(), testBench.getCurrentTime()),
      new DomainMessage(id2, 2, new TestEvent2(), testBench.getCurrentTime()),
      new DomainMessage(id2, 3, new TestEvent(), testBench.getCurrentTime()),
    ]);
  });

  it('Different aggregates', async () => {

    class TestAggregate2 extends EventSourcedAggregateRoot {

    }

    const testBench = new EventSourcingTestBench();

    const id1 = Identity.create();
    const id2 = Identity.create();

    await testBench
      .given(id1, TestAggregate, [
        new TestEvent(),
      ])
      .given(id2, TestAggregate2, [
        new TestEvent(),
      ]);

    const context1 = testBench.getAggregateTestContext(TestAggregate);
    const stream1 = await context1.getEventStore().loadAll();
    const messages1 = await stream1.toArray().toPromise();

    expect(messages1).toEqual([
      new DomainMessage(id1, 0, new TestEvent(), testBench.getCurrentTime()),
    ]);

    const context2 = testBench.getAggregateTestContext(TestAggregate2);
    const stream2 = await context2.getEventStore().loadAll();
    const messages2 = await stream2.toArray().toPromise();

    expect(messages2).toEqual([
      new DomainMessage(id2, 0, new TestEvent(), testBench.getCurrentTime()),
    ]);
  });

});

it('givenCurrentTime should assign the time', () => {
  const testBench = new EventSourcingTestBench();
  const time = new Date();
  testBench.givenCurrentTime(time);
  expect(time).toBe(time);
});

it('whenTimeChanges should assign the time', () => {
  const testBench = new EventSourcingTestBench();
  const time = new Date();
  testBench.whenTimeChanges(time);
  expect(time).toBe(time);
});

it('whenCommands should dispatch commands on command bus', async () => {
  class TestCommand implements Command {

  }

  const testBench = new EventSourcingTestBench();

  testBench.commandBus.dispatch = jest.fn() as any;

  const command1 = new TestCommand();
  const command2 = new TestCommand();
  await testBench.whenCommands([command1, command2]);

  expect(testBench.commandBus.dispatch).toBeCalledWith(command1);
  expect(testBench.commandBus.dispatch).toBeCalledWith(command2);
});

describe('whenDomainMessagesHappened should dispatch messages on event bus', () => {
  it('by array', async () => {
    class TestEvent implements DomainEvent {

    }

    const testBench = new EventSourcingTestBench();

    const id = Identity.create();

    const mock = jest.fn();
    testBench.eventBus.publish = mock;

    await testBench.whenDomainMessagesHappened([
      new DomainMessage(id, 0, new TestEvent(), testBench.getCurrentTime()),
      new DomainMessage(id, 1, new TestEvent(), testBench.getCurrentTime()),
    ]);

    const stream: DomainEventStream = mock.mock.calls[0][0];

    expect(await stream.toArray().toPromise()).toEqual([
      new DomainMessage(id, 0, new TestEvent(), testBench.getCurrentTime()),
      new DomainMessage(id, 1, new TestEvent(), testBench.getCurrentTime()),
    ]);
  });

  it('by stream', async () => {
    const testBench = new EventSourcingTestBench();

    const mock = jest.fn();
    testBench.eventBus.publish = mock;

    const stub = SimpleDomainEventStream.of([]);

    await testBench.whenDomainMessagesHappened(stub);

    const stream: DomainEventStream = mock.mock.calls[0][0];
    expect(stream).toBe(stub);
  });
});

it('whenEventsHappened should create messages and call whenDomainMessagesHappened', () => {
  class TestEvent implements DomainEvent {

  }

  const testBench = new EventSourcingTestBench();
  testBench.domainMessageFactory.createDomainMessages = jest.fn().mockReturnValueOnce('stub stream');
  testBench.whenDomainMessagesHappened = jest.fn();
  const id = Identity.create();
  testBench.whenEventsHappened(id, [new TestEvent()]);

  expect(testBench.domainMessageFactory.createDomainMessages).toBeCalledWith(id, [new TestEvent()]);
  expect(testBench.whenDomainMessagesHappened).toBeCalledWith('stub stream');
});

describe('Should validate date', () => {
  /* tslint:disable:no-console */
  // disable warn message of moment.
  const warn = console.warn;
  console.warn = () => { /* noop */ };
  afterAll(() => {
    console.warn = warn;
  });

  it('By constructor', () => {
    expect(() => {
      return new EventSourcingTestBench('no valid date');
    }).toThrow();
  });

  it('By givenCurrentTime', () => {
    const testBench = new EventSourcingTestBench();
    expect(() => {
      testBench.givenCurrentTime('not valid');
    }).toThrow();
  });

  it('By whenTimeChanges', () => {
    const testBench = new EventSourcingTestBench();
    expect(() => {
      testBench.whenTimeChanges('not valid');
    }).toThrow();
  });

});
