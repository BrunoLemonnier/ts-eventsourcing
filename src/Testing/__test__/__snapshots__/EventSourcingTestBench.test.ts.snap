// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logging Logs messages 1`] = `
Array [
  "src/Testing/__test__/EventSourcingTestBench.test.ts:699:8    givenReadModelRepository",
  "src/Testing/__test__/EventSourcingTestBench.test.ts:702:8    thenAssert",
  "    src/Testing/EventSourcingTestBench.ts:656:18                 thenWaitUntilProcessed",
  "    src/Testing/__test__/EventSourcingTestBench.test.ts:703:25   thenModelsShouldMatch",
  "    src/Testing/EventSourcingTestBench.ts:537:18                 thenWaitUntilProcessed",
]
`;

exports[`Logging The repository class arguments should be seen in the logs 1`] = `
Array [
  "src/Testing/__test__/EventSourcingTestBench.test.ts:749:8    givenReadModelRepository",
  "src/Testing/__test__/EventSourcingTestBench.test.ts:750:8    givenQueryHandler",
  "    Created class TestQueryHandlerWithRepository with arguments:",
  "           ProductAggregate --> EventSourcingRepository",
  "              UserAggregate --> EventSourcingRepository",
  "              UserReadModel --> InMemoryRepository",
  "               MyRepository --> MyRepository",
]
`;

exports[`Logging defaults Default logs to process 1`] = `
Array [
  "src/Testing/__test__/EventSourcingTestBench.test.ts:768:10   givenReadModelRepository 
",
  "src/Testing/__test__/EventSourcingTestBench.test.ts:771:10   thenAssert 
",
  "    src/Testing/EventSourcingTestBench.ts:656:18                 thenWaitUntilProcessed 
",
  "    src/Testing/__test__/EventSourcingTestBench.test.ts:772:27   thenModelsShouldMatch 
",
  "    src/Testing/EventSourcingTestBench.ts:537:18                 thenWaitUntilProcessed 
",
]
`;

exports[`givenCommandHandler should be able to give own repository by default constructor interface 1`] = `
TestRepository {
  "aggregateFactory": SimpleEventSourcedAggregateFactory {
    "aggregate": [Function],
  },
  "eventBus": RecordDomainEventBusDecorator {
    "bus": AsynchronousDomainEventBus {
      "activeStreamSubscription": null,
      "errorHandler": [Function],
      "eventHandlersMappedByEvent": Object {},
      "isProcessing": false,
      "isProcessingSubject": Subject {
        "_isScalar": false,
        "closed": false,
        "hasError": false,
        "isStopped": false,
        "observers": Array [],
        "thrownError": null,
      },
      "onComplete": [Function],
      "queue": Array [],
    },
    "messages": Array [],
  },
  "eventStore": InMemoryEventStore {
    "events": Object {},
  },
  "streamDecorator": AggregateDomainEventStreamDecorator {
    "decorators": Array [
      OverrideDateDomainEventStreamDecorator {
        "getDate": [Function],
      },
    ],
  },
}
`;
